core/slice.go

```diff
@@ -0,0 +1,484 @@
+		if order < types.QuaiNetworkContext {
+			status := sl.hc.GetHeaderByHash(terminalHeader.Hash())
+			fmt.Println("terminal Header status", status)
+			if status != nil {
+				if prevTerminalHeader.Hash() != header.Hash() {
+					return nil, errors.New("subordinate terminus mismatch")
+				}
+				return terminalHeader, nil
+			}
+		} else if order == types.QuaiNetworkContext {
+			return terminalHeader, err
+		}
+
+		prevTerminalHeader = terminalHeader
+	}
+}
+
+// PreviousCoincidentOnPath searches the path for a block of specified order in the specified slice
+//     *slice - The zone location which defines the slice in which we are validating
+//     *order - The order of the conincidence that is desired
+//     *path - Search among ancestors of this path in the specified slice
+func (sl *Slice) PreviousCoincidentOnPath(header *types.Header, slice []byte, order int, fullSliceEqual bool) (*types.Header, error) {
+
+	if header.Number[types.QuaiNetworkContext].Cmp(big.NewInt(0)) == 0 {
+		return sl.hc.GetHeaderByHash(sl.hc.Config().GenesisHashes[0]), nil
+	}
+
+	if err := sl.hc.CheckContext(order); err != nil {
+		return nil, err
+	}
+	if err := sl.hc.CheckLocationRange(slice); err != nil {
+		return nil, err
+	}
+
+	for {
+		// If block header is Genesis return it as coincident
+		if header.Number[types.QuaiNetworkContext].Cmp(big.NewInt(1)) == 0 {
+			return sl.hc.GetHeaderByHash(sl.hc.Config().GenesisHashes[0]), nil
+		}
+
+		// Get previous header on local chain by hash
+		prevHeader := sl.hc.GetHeaderByHash(header.ParentHash[types.QuaiNetworkContext])
+		if prevHeader == nil {
+			return nil, consensus.ErrSliceNotSynced
+		}
+		// Increment previous header
+		header = prevHeader
+
+		// Find the order of the header
+		difficultyOrder, err := sl.GetDifficultyOrder(header)
+		if err != nil {
+			return nil, err
+		}
+
+		// If we have reached a coincident block of desired order in our desired slice
+		var equal bool
+		if fullSliceEqual {
+			equal = bytes.Equal(header.Location, slice)
+		} else {
+			equal = header.Location[0] == slice[0]
+		}
+		if equal && difficultyOrder <= order {
+			return header, nil
+		}
+	}
+}
+
+// HLCR does hierarchical comparison of two difficulty tuples and returns true if second tuple is greater than the first
+func (sl *Slice) HLCR(localDifficulties []*big.Int, externDifficulties []*big.Int) bool {
+	if len(externDifficulties) == 0 || len(localDifficulties) == 0 {
+		return false
+	}
+	if localDifficulties[0].Cmp(externDifficulties[0]) < 0 {
+		return true
+	} else if localDifficulties[0].Cmp(externDifficulties[0]) > 0 {
+		return false
+	}
+	if localDifficulties[1].Cmp(externDifficulties[1]) < 0 {
+		return true
+	} else if localDifficulties[1].Cmp(externDifficulties[1]) > 0 {
+		return false
+	}
+	if localDifficulties[2].Cmp(externDifficulties[2]) < 0 {
+		return true
+	} else if localDifficulties[2].Cmp(externDifficulties[2]) > 0 {
+		return false
+	}
```

Comment:
Instead of a cascading decision tree, it might be better to loop as such:
```
for ctx := 0; ctx < len(localDifficulties); ctx++ {
	if localDifficulties[ctx].Cmp(externDifficulties[ctx]) < 0 {
		return true
	}
}
return false
```

There are a few things that will be easier (if we decide to do them):
* if we decide to add contextual selfish mining mitigation
* if we decide to expand or reduce the hierarchy ever